File: common/src/main/java/link/e4steam/steam/SteamLifecycle.java
package link.e4steam.steam;

import java.io.IOException;

/** Restartable ownership of the process-global Steam API. */
final class SteamLifecycle implements AutoCloseable {
    private final SteamApi api;
    private boolean librariesLoaded;
    private boolean initialized;

    SteamLifecycle(SteamApi api) {
        this.api = api;
    }

    void start() throws IOException {
        if (initialized) {
            return;
        }
        if (!librariesLoaded) {
            SteamNativeLibraryLoader loader = new SteamNativeLibraryLoader();
            if (!api.loadLibraries(loader)) {
                throw new IOException(
                        "Could not load Steam native libraries: " + loader.failureDescription(),
                        loader.failureCause()
                );
            }
            librariesLoaded = true;
        }
        try {
            if (!api.init()) {
                throw new IOException("SteamAPI_Init failed. Start Steam and sign in before launching Minecraft");
            }
        } catch (IOException exception) {
            throw exception;
        } catch (Exception exception) {
            throw new IOException("SteamAPI_Init failed: " + exception.getMessage(), exception);
        }
        initialized = true;
        if (!api.isSteamRunning()) {
            close();
            throw new IOException("Steam is not running or the current user is not signed in");
        }
    }

    void runCallbacks() {
        if (!initialized) {
            throw new IllegalStateException("Steam lifecycle is not running");
        }
        api.runCallbacks();
    }

    boolean isRunning() {
        return initialized && api.isSteamRunning();
    }

    @Override
    public void close() {
        if (initialized) {
            initialized = false;
            api.shutdown();
        }
    }
}

File: common/src/main/java/link/e4steam/steam/SteamLifecycle.kt
package link.e4steam.steam

import java.io.IOException
import java.nio.file.Path

/** Restartable ownership of the process-global Steam API. */
class SteamLifecycle(private val api: SteamApi) : AutoCloseable {
    private var nativeLoader: SteamNativeLibraryLoader? = null
    private var librariesLoaded = false
    private var initialized = false

    @Throws(IOException::class)
    fun start() {
        if (initialized) {
            return
        }
        if (!librariesLoaded) {
            val loader = SteamNativeLibraryLoader()
            nativeLoader = loader
            if (!api.loadLibraries(loader)) {
                throw IOException(
                    "Could not load Steam native libraries: " + loader.failureDescription(),
                    loader.failureCause()
                )
            }
            librariesLoaded = true
        }
        try {
            if (!api.init()) {
                throw IOException("SteamAPI_Init failed. Start Steam and sign in before launching Minecraft")
            }
        } catch (exception: IOException) {
            throw exception
        } catch (exception: Exception) {
            throw IOException("SteamAPI_Init failed: " + exception.message, exception)
        }
        initialized = true
        if (!api.isSteamRunning()) {
            close()
            throw IOException("Steam is not running or the current user is not signed in")
        }
    }

    fun runCallbacks() {
        if (!initialized) {
            throw IllegalStateException("Steam lifecycle is not running")
        }
        api.runCallbacks()
    }

    fun isRunning(): Boolean = initialized && api.isSteamRunning()

    fun steamApiPath(): Path? {
        if (!initialized || nativeLoader == null) {
            throw IllegalStateException("Steam lifecycle is not running")
        }
        return nativeLoader!!.steamApiPath()
    }

    override fun close() {
        if (initialized) {
            initialized = false
            api.shutdown()
        }
    }
}
